### PR TITLE
#255 Fix for UI bug: dropdown menus on request detail page don't update when adding a request or adding/removing elements

### LINKED ISSUE #255: Creating a new request from the Request Details page does not refresh menu
# Description

image

The "Create request" button used in the Request Detail page to add elements to other requests will create a new request and map the data element to that new request, but the list of requests created does not refresh to show the new request.

# Steps to reproduce

1. View any data request detail, either unsent or submitted.
2. Click the "+" button next to a element, class or schema
3. Click the "Create new request" button in the popup menu
4. Enter a name for the request that has not been used before, then click "Create request"
5. Let the operation complete. The success dialog should appear and you should the continue browsing.
6. Click the same "+" button again and check the tick list. The new request you created is not there.
7. Do a hard refresh of the browser page and check the "+" menu again. The new request is now there and ticked.

# Expected behavior

After the operation has completed successfully, the menu should automatically refresh without the user having to do anything.

# Additional context

A broadcast event called `data-request-added` is sent globally via the `DataRequestsService` when the request is created and updated. I think this needs to be listened to in the Request Detail page to update the list of request intersections, which seems to be missing at the moment.

### PR DESCRIPTION
**Pull request by greg-at-occ**
Read the issue for full description. The issue concerns only _new requests_ but this PR also fixes a similar issue when elements are added to/removed from other requests and the tick box states on the other drop-down menus would be left out-of-date.

For new data requests, we set up a handler for the data-request-added signal, which re-calculates the intersections and issues the updated result via the data-intersections-refreshed signal. This was chosen despite the intersections already propagating down to the child elements because of some performance concerns around using ngOnChanges and because the handling for  data-intersections-refreshed was already present in data-element-in-request.

For data element insert/remove changes, we piggyback on the existing requestAddDelete Output where a handler was already present for a less general case. Renaming and extending this resolves the issue.

Closes #255

### CODE REVIEW

File: src/app/pages/my-request-detail/my-request-detail.component.ts [hunk 1]
```diff
@@ -112,6 +114,11 @@ export class MyRequestDetailComponent implements OnInit {
 
   selectAllElements = false;
 
+  /**
+   * Signal to attach to subscriptions to trigger when they should be unsubscribed.
+   */
+  private unsubscribe$ = new Subject<void>();
```
Comment (CristianGimenez) [inline src/app/pages/my-request-detail/my-request-detail.component.ts:120]: Do we need new tests on this component to test this works as expected (which it seems it does) but mostly to ensure we don't break anything in the future?

**Review by CristianGimenez — CHANGES_REQUESTED** [review]
Generally it looks great. All tests pass. Pr checks all good as well. Very neat first PR, it seemed like a dodgy issue for someone new on the project and you handled gratefully, well done!

I've made a couple of comments:

- [ ] Include tests (I'm happy to help, Jest can be confusing) 
- [ ] Small change on code comment
- [ ] Suggested small refactor to avoid code duplication (optional)

File: src/app/pages/my-request-detail/my-request-detail.component.ts [hunk 2]
```diff
@@ -281,9 +294,14 @@ export class MyRequestDetailComponent implements OnInit {
   }
 
   // listens for external update to the request and refreshes if so
-  handlePossibleSelfDelete(event: RequestElementAddDeleteEvent) {
+  handleRequestElementsChange(event: RequestElementAddDeleteEvent) {
     if (event.dataModel?.id === this.request?.id) {
       this.setRequest(this.request);
+    } else {
+      this.loadIntersections(this.dataSchemas).subscribe((intersections) => {
```
[the remark this answers is not in the document]
Comment (greg-at-occ) [reply src/app/pages/my-request-detail/my-request-detail.component.ts]: Done

File: src/app/pages/my-request-detail/my-request-detail.component.ts [hunk 3]
```diff
@@ -112,6 +114,11 @@ export class MyRequestDetailComponent implements OnInit {
 
   selectAllElements = false;
 
+  /**
+   * Signal to attach to subscriptions to trigger when they should be unsubscribed.
```
[the remark this answers is not in the document]
Comment (greg-at-occ) [reply src/app/pages/my-request-detail/my-request-detail.component.ts]: Done

File: src/app/pages/my-request-detail/my-request-detail.component.ts (hunk 1 above)
Comment (greg-at-occ) [reply src/app/pages/my-request-detail/my-request-detail.component.ts:120]: Started but abandoned to allow us to close down some in-flight branches on this page

**Review by CristianGimenez — APPROVED** [review]
Thanks for addressing the comments. I'm happy with the changes, approving PR.

**Review by pjmonks — APPROVED** [review]
@CristianGimenez has already approved this, just re-approving after rebase

**Pull request merged**